TACC/Core-CMS
# Quick: Support Non-Bootstrap Link & Picture Plugins

# Overview

Restore support for Generic plugins that `djangocms_bootstrap4` removes.

<details><summary>Why?</summary>

Most of the time, the extra features of Bootstrap4's link and image are:
1. unused by users
2. undesired by designers
3. add unnecessary complexity to development of TACC plugins that need simple image and link feature

</details>

# Changes

- Register new TACC plugin to re-register `LinkPlugin` (if Bootstrap's respective plugin is present).
- Register new TACC plugin to re-register `PicturePluign` (if Bootstrap's respective version is present).

# Screenshots

__Note__: I propose we mark the relevant Bootstrap plugins as deprecated.

<details>

Though Bootstrap plugins are still available (to support existing usage), they will __not__ be allowed to be nested within TACC plugins. Also, TACC plugins will extend Generic plugins, not Bootstrap plugins.

</details>
<details><summary>Legend</summary>

- Blue plugins are for links.
- Yellow plugins are for images.
- Outlined Bootstrap plugins are undesirable.
- Outlined Generic plugins are undesirable.

</details>

| Description | Image 1 | Image 2 |
| --- | --- | --- |
| | Link | Image |
| Add Plugins | ![Link Add](https://user-images.githubusercontent.com/62723358/131549805-b1d75cc3-c62c-4f06-ae9e-2de942255bca.png) | ![Image Add](https://user-images.githubusercontent.com/62723358/131549817-9f380cf5-00a7-48c6-8732-bf66370aaaeb.png) |
| Generic Plugin Form | ![Generic Link Form](https://user-images.githubusercontent.com/62723358/131549237-baedaa2a-5b69-496a-bed3-3e6ea41ff7d2.png) | ![Generic Image Form](https://user-images.githubusercontent.com/62723358/131549329-5ae323c1-48d0-4a86-b8db-e984d0d6ae8b.png) |
| Bootstrap Plugin Form | ![Bootstrap Link Form](https://user-images.githubusercontent.com/62723358/131549236-e37f0780-d433-4873-8a38-e64e54b2de99.png) | ![Bootstrap Image Form](https://user-images.githubusercontent.com/62723358/131549353-3c655b1b-c8a5-40c0-8608-53fb22bb0f82.png) |
| | Not Within "Text" Plugin | Within "Text" Plugin |
| Render Plugins | ![Bootstrap   Generic - Render - Not Within Text](https://user-images.githubusercontent.com/62723358/131545561-3aa36821-e84d-4551-af91-2fb52d263799.jpeg) | ![Bootstrap   Generic - Render - Within Text](https://user-images.githubusercontent.com/62723358/131545565-d1362d34-1504-44c6-bb5a-66bcd097411d.png) |
| Plugin Structure | ![Bootstrap   Generic - Structure - Not Within Text](https://user-images.githubusercontent.com/62723358/131550287-1124662b-7a20-4867-884a-e587351ae66b.png) | ![Bootstrap   Generic - Structure - Within Text](https://user-images.githubusercontent.com/62723358/131550290-f39f9249-a7a1-4e18-9fc3-d320b91158da.png) |

# Testing

You may test locally or on servers that have deployed this code.

> __Last Recorded Deploys__
> - [Frontera (dev)](https://dev.fronteraweb.tacc.utexas.edu/) was [Build #454 (Aug 9, 2021 1:02:37 PM)](https://jenkins01.tacc.utexas.edu/job/Core_Portal_Deploy/454/) via `site/main/frontera` (which has this branch and more)
> - [Core (dev)](https://dev.cep.tacc.utexas.edu/) was [Build #453 (Aug 9, 2021 1:04:25 PM)](https://jenkins01.tacc.utexas.edu/job/Core_Portal_Deploy/453/) via `site/main/frontera` (which has this branch and more)

1. Ensure that both "Generic" plugins, "Link" and "Image", are:
	- available to add
	- functional
	- __not__ tagged as needing attention
	- __not__ described as deprecated
2. Ensure that both "Bootstrap4" plugins, "Link / Button" and "Picture / Image", are:
	- available to add
	- functional
	- tagged as needing attention
	- described as deprecated

# Notes

## Known Issues

1. <details>
	<summary>Should We Prevent New Instances of Bootstrap Plugins, but Support Existing Ones?</summary>

	__Maybe.__ It depends on whether there are valid use cases for the plugins extra features that TACC cannot or will not provide.

	</details>

2. <details>
	<summary>How Would We Prevent New Instances of Bootstrap Plugins, but Support Existing Ones?</summary>

	Tasks:
	- Hide plugins from plugin add list from non-admin users (so admins retain full control).
	- But show plugins in structure list (so they can be edited if already in use).

	Ideas:
	- CSS:
    	```css
    	/* HELP: How to limit this to non-admin users? */
    	div.cms .cms-plugin-picker .cms-submenu-item a[data-rel="add"][href="TaccsiteNewsArticleListPlugin"] {
        	display: none;
    	}
    ```

	</details>
